### ISSUE TITLE
v2.2.3 matplotlib.pyplot.plot() creates empty plot. savefig() works as expected. 

### ISSUE THREAD

**Issue by mmenschig** [matplotlib/matplotlib#12074]
### Bug report

**Bug summary**

When using the latest stable version of matplotlib (v2.2.3), running plt.plot() only displays a blank figure, regardless of chosen backend. However, the .savefig() method correctly stores an image of the plot.

Switching to an older version (2.0.2) resolves the issue.

**Code for reproduction**

```python
import matplotlib as mpl
mpl.use("TKAgg")

import matplotlib.pyplot as plt

plt.plot([1, 2, 3, 4])
plt.ylabel('some numbers')
plt.savefig('testfigure.png', dpi=100)
plt.show(block=True)
```

**Actual outcome**

![Image](https://i.imgur.com/fo24aqO.png)

No code errors.

**Expected outcome**

![Image](https://i.imgur.com/wmerK0B.png)

**Matplotlib version**
2.2.3 - works on 2.0.2
  * Operating system: MacOS Mojave (10.14 Beta (18A384a))
  * Matplotlib version: 2.2.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: 3.7
  * Jupyter version (if applicable): N/A
  * Other libraries: N/A

Installed via `pip install matplotlib`

**Comment by mmenschig** [comment]
Further testing shows that the bug does not reproduce using versions 2.1.2.

**Comment by jklymak** [comment]
I cannot reproduce using master or v2.2.3.  My guess would be pip didn't install something for you properly...

**Comment by mmenschig** [comment]
@jklymak running macOS?

**Comment by jklymak** [comment]
Yes, but I install via conda.

**Comment by mmenschig** [comment]
@jklymak Interesting. Just installed anaconda (I don't usually use it, so may not keep it) and created a new env and installed matplotlib, ran the code and voilà, the figure shows the data as expected.

Doesn't seem to be an issue with matplotlib at all after all. Any thoughts on what to try next? Think this is an issue with pip?

**Comment by phobson** [comment]
How fresh is your pip-based environment that produces this issue? Can you provide steps to recreate the environment?

**Comment by tacaswell** [comment]
This may be an issue with tk.  We changed details about how exactly the c-extension we need for tk is built / linked in 2.2 (to support pypy!). 

What version of osx?  This may also be an issue with the wheels and conflicts between the tk you have, the tk they were built with, and the version of OSX they were built with vs the version of OSX you have.

**Comment by jklymak** [comment]
Closing as I don't think this is us.  I think either Tk needs to be updated or there is an issue w/ OSX

**Comment by mmenschig** [comment]
Thank you @jklymak 

I'll be revisiting this issue when the current beta version of macOS gets released and provide an update then.

**Comment by shamoons** [comment]
Can we reopen? I'm having this issue with 10.14.3

**Comment by jklymak** [comment]
@shamoons can you re-open a full issue, including versions, what you did to install, and what you have tried to fix the problem.  As noted above, this is usually an installation problem with an old library lying around from a previous install somehow not being cleared out.  Thanks...

**Comment by AlphaSRE** [comment]
> The issue appears with macOS Mojave (10.14 Beta (18A384a))
> Just tested this with macOS Sierra (10.12.6) and the figure is displayed correctly.

I have the same problem.

* Matplotlib version:3.0.2

**Comment by jeffreywildman** [comment]
Experiencing the same problem. Steps I used to reproduce:

```bash
brew install pyenv pipenv
pipenv install matplotlib
pipenv run python -c "import matplotlib; matplotlib.use('tkagg'); import matplotlib.pyplot as plt; plt.plot([1,2,3,4]); plt.savefig('test.png'); plt.show()"
```

Version information of relevant tools:

```bash
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.14.3
BuildVersion:	18D109
$ pipenv --version
pipenv, version 2018.11.26
$ pyenv --version
pyenv 1.2.9
$ pipenv run python --version
Python 3.7.2
$ pipenv run python -c "import matplotlib; print(matplotlib.__version__)"
3.0.2
```

**Comment by tacaswell** [comment]
Is `pyenv` the one that builds python for you? If so, do you have the correct tk headers installed at the time you built python?

Are you pulling the wheels or the tarball?

Can you reproduce this issue without `pyenv` / `pipenv`?  They are doing enough 'helpful' things for you I do not have a good sense of what they are _actually_ doing.

**Comment by jeffreywildman** [comment]
Removal of `pyenv` from above also resulted in display of a blank image.
However, additionally removing `pipenv` resulted in a successful plot to screen:

```bash
python3 -m venv tmp
./tmp/bin/pip3 install matplotlib
./tmp/bin/python3 -c "import matplotlib; matplotlib.use('tkagg'); import matplotlib.pyplot as plt; plt.plot([1,2,3,4]); plt.savefig('test.png'); plt.show()"
```

I'm not yet sure whether `pipenv` is providing its own python3 or pulling the available `python3` from `brew`, or something else.

**Comment by tacaswell** [comment]
> However, additionally removing `pipenv` resulted in a successful plot to screen:

Should this discussion be moved to `pipenv`?  I think this is an installation issue so we need to sort out what pipenv is doing under the hood and some details about what versions of tk are used for both the cpython binary and the wheel.

**Comment by jeffreywildman** [comment]
I've narrowed this down to an issue with `virtualenv`, which `pipenv` uses to manage virtual environments. 

Working using `venv`:

```bash
python3 -m venv tmp
./tmp/bin/pip3 install matplotlib
./tmp/bin/python3 -c "import matplotlib; matplotlib.use('tkagg'); import matplotlib.pyplot as plt; plt.plot([1,2,3,4]); plt.savefig('test.png'); plt.show()"
```

Non-working using `virtualenv`:

```bash
pip3 install virtualenv
virtualenv tmp
./tmp/bin/pip3 install matplotlib
./tmp/bin/python3 -c "import matplotlib; matplotlib.use('tkagg'); import matplotlib.pyplot as plt; plt.plot([1,2,3,4]); plt.savefig('test.png'); plt.show()"
```

Some FAQs that I'm still processing:
https://matplotlib.org/faq/virtualenv_faq.html
https://matplotlib.org/faq/osx_framework.html

Related reading from pipenv:
https://github.com/pypa/pipenv/issues/15

**Comment by CountingClouds** [comment]
Hi @jeffreywildman and all - please share if you learn how to resolve this

**Comment by markserrano915** [comment]
I am using ```pyenv```

```
pyenv versions
  system
* 3.7.2 (set by /Users/mserrano/.pyenv/version)
```

If I use 3.7.2 via ```pyenv global 3.7.2```, I get a white screen whenever I try to plot.

But if I use the system ```pyenv global system```, I get the plot correctly.

**Comment by Popiko** [comment]
I am using virtualenv with python 2.7.10 in macOS 10.14.3 and get the same problem.
My code
```python
import sys
import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
import numpy as np

data = np.random.rand(100)

fig, (ax1, ax2) = plt.subplots(1, 2)
ax1.plot(data)
ax2.plot(data)
ax1.plot(data+1)

plt.show()

```
And I got the blank plot using Matplotlib2.2.3
<img width="638" alt="issue1" src="https://user-images.githubusercontent.com/48267604/56450497-4d33f280-6359-11e9-80a8-1683e21d6873.png">
Without warning or error

After I switch to Matplotlib 2.1.2 I can get the correct plot
<img width="638" alt="issue2" src="https://user-images.githubusercontent.com/48267604/56450529-6dfc4800-6359-11e9-90ec-4c02a51ffb5a.png">

**Comment by tacaswell** [comment]
Suspect that https://github.com/matplotlib/matplotlib/issues/13988 is related.

There is clearly something wrong with how we are handling the tk main loop or window on _some_ systems / enviroments.

To sort this out we need someone who has a broken system to dig into the tk side of things...

**Comment by alexattia** [comment]
Same issue with Mojave 10.14.4 and Python 3.

**Comment by sannecottaar** [comment]
I'm having the same issue with Mojave 10.14.2, Python 3 in a conda environment. Going back to older matplotlib version (back to 2.0.2) did not fix the issue for me.

**Comment by detailyang** [comment]
it works fallback to 2.1.2

**Comment by sannecottaar** [comment]
Unfortunately, I had not luck in fixing it that way.

**Comment by LLansing** [comment]
I encountered this issue while using pyenv on 10.14.4 Mojave macOS. Falling back to an older version of matplotlib did not work for me, but uninstalling pyenv entirely did.

**Comment by efiring** [comment]
I'm seeing no end of Tk problems with conda environments and a variety of mpl versions on OSX.  There are definitely event-handling problems.  In the simplest case such as the original test shown here, at the moment the main glitch is that the toolbar buttons do not reliably respond to clicks.  Often it is necessary to click in the axes frame and then go back to a toolbar button to get the click to register.  The button gives a visual feedback of the click, but has no other effect.  The cursor readout is perfectly responsive, though.
As another example  of a glitch, clicking on the subplots-adjuster button brings up a black window that isn't drawn until it is resized.
@jklymak since you are also using conda, would you give this a quick try?  Make a minimal plot and try various toggling of the navbar buttons.  I'm on Mojave now, but I think I have seen the erratic button response with a conda environment on a Sierra machine as well.  I'm trying mpl 2.2.2 and 3.1.0.

Ha!  I just now tried 3.1.0+1017.g4b50ce6cc and got the infamous blank white window!
Conda tk version is 8.6.9.

**Comment by efiring** [comment]
At the moment, at least, I can reliably reproduce the white window with master, Mojave, conda py3.6, conda tk 8.6.9, running the following script from the command line:
```python
import matplotlib
matplotlib.use('tkagg')
import matplotlib.pyplot as plt
plt.plot([1,2,3])
plt.show()
```
I have added print statements to verify that `draw` is being called, and that `blit` is getting a plausible array generated by agg.  It seems that the last part of it, presumably the C++ function being called, simply is not getting that image to the screen with my particular combination of OS and libraries.
```python
    _tkagg.blit(
        photoimage.tk.interpaddr(), str(photoimage), dataptr, offsets, bboxptr)
```
@anntzer, do you have any ideas for *simple* additional debugging I might be able to do to figure out what is going wrong here?

**Comment by efiring** [comment]
One more point: upon executing the script the window is initially white.  Resizing it turns it black.

**Comment by anntzer** [comment]
No real ideas here, but can you summarize the behavior on whatever systems, both before 3.0 and after? (The only substantial recent change I can think of is #10680, which went in in 3.0.)  I'm a bit lost with who sees what behavior now.

**Comment by efiring** [comment]
You can skip to the bottom if you like; the problem seems to be conda-forge (and maybe other) tk builds and versions.  Summary: 8.5.19 looks good, 8.6.9 is flaky from one build to the next.

In the "mpl1" py37 conda environment that I use for working on mpl, always using `pip install -e .` to install it, I get the white screen with versions 1.5.3, 2.2.4, 3.0.0, and master.   I also tried 2.2.0 installed without the `-e` and got the same result.

In my relatively old "py36" conda environment with mpl v2.2.2 that I use day in and day out, and in a newer "py37" environment with mpl v3.1.0 I get a normal-looking plot, but the response to the navigation toolbar buttons is erratic.

That makes it sound like there is a difference based on whether mpl is installed as a conda package or from source on git via pip, *but* ... on a colleague's machine (Mojave) with everything from a fresh py37 conda environment, the white screen appears.

And now, when I make a fresh py37 environment from scratch:
```
conda create -n mpl2 python=3.7 matplotlib tk
```
and run the test, I get the white screen.

So it seems that the problem is not at all dependent on mpl version, and not really dependent on installation method, either.  Instead, it might depend on the exact tk build:

My original "py37" environment (yielding a plot) had tk 8.6.9-ha92aebf_0 from conda-forge.  Now I updated that to match the fresh environment ("mpl2") version, 8.6.9-h2573ce8_1002, and I get the white screen.  Switching back to the earlier tk build, I get the plot again.

When I downgrade tk in "mpl2" to 8.5.19, it downgrades a bunch of other things including python to 3.6; but the test yields a window, and the buttons work!

@ocefpaf it looks like there might be something very fiddly with tk versions and builds for OSX in conda-forge.

**Comment by anntzer** [comment]
Given the dependency on various tk versions that I cannot test myself, there isn't much I can do about it :(
The Tcl/Tk changelog for 8.6 (https://wiki.tcl-lang.org/page/Changes+in+Tcl%2FTk+8.6) does mention "Tk for OS X/Aqua now built on Cocoa libraries."; not sure how relevant this is but may be worth investigating?

**Comment by janosh** [comment]
This has been bugging me for quite a while now. From skimming this issue, it seems it's still unclear where this problem originates?

**Comment by efiring** [comment]
Yes, it's unclear. It would be great to have someone with expertise in all 3 of Tk, OSX, and mpl to track down the problem and solve it.

**Comment by ocefpaf** [comment]
> @ocefpaf it looks like there might be something very fiddly with tk versions and builds for OSX in conda-forge.

Opened https://github.com/conda-forge/tk-feedstock/issues/36 to track this.

**Comment by AhadCove** [comment]
What worked for me was literally just removing the backend line inside `~/.matplotlib/matplotlibrc`
I just left the file blank and everything worked perfectly after that

**Comment by tacaswell** [comment]
This looks like it is issues with how tk is built and its interaction with OSX , closing as there is nothing we can do about it at the Matplotlib level.

**Comment by vboikotriller** [comment]
I have the same issue on Linux Debian.
matplotlib==3.3.4
Pillow==8.1.0

python3-tk 3.9.1-2

"""
import matplotlib.pyplot as plt
import numpy as np
plt.title("Title")
plt.plot(np.linspace(-1, 1, 50), np.linspace(-2, 2, 50))
"""

If I take title out, it works. Why matplotlib can't display titles ?

**Comment by jklymak** [comment]
If you have checked your install please open a new issue.

**Comment by dstansby** [comment]
Using `tk                        8.6.12               he1e0b03_0    conda-forge` (the latest version) this seems to be working fine for me, so I'm going to close this. If this is still an issue for anyone please comment and I will re-open!
